### PR TITLE
Adapt app to new error handler

### PR DESCRIPTION
**Pull request by pbochis**
- make base provider behavior parse errors
- adapt iron-ajax wrapper(coduno-iron-ajax) to be used for post/put requests with error handling incorporated
- replace iron-ajax with coduno-iron-ajax where approprivate
- rework error page to suit multiple errors(500, 403, 404)

### CODE REVIEW

File: app/elements/admin/import-catcoder-input.html [hunk 1]
```diff
@@ -2,55 +2,66 @@
...
+			paper-button {
+				margin: 8px 4px ;
+			}
+		</style>
+		<coduno-iron-ajax
 			id="postGame"
 			method="POST"
-			with-credentials="true"
+			body="{{body}}"
+			suffix="/contestgameupload"
```
Comment (victorbalan) [inline app/elements/admin/import-catcoder-input.html]: rename to path

File: app/elements/common/coduno-iron-ajax.html [hunk 2]
```diff
@@ -45,30 +43,34 @@
...
+					util.error(this.error);
+					return;
+				}
+				if (this.error.status === 401){
+					page('/login');
+				}
+				this.fire('error', e);
+			},
+			_buildUrl: function(){
+				return util.build(this.suffix);
```
Comment (victorbalan) [inline app/elements/common/coduno-iron-ajax.html]: don`t use util. Please take a look at providers and try to do something similar

File: app/elements/common/coduno-iron-ajax.html [hunk 3]
```diff
@@ -3,15 +3,14 @@
...
 	<template>
 		<iron-ajax
-		  id="ajaxRequest"
-		  method="{{method}}"
-		  content-type="{{contentType}}"
-		  handle-as="{{handleAs}}"
-		  last-response="{{lastResponse}}"
-		  headers="{{headers}}"
-		  url="{{url}}"
+			id="ajaxRequest"
```
Comment (victorbalan) [inline app/elements/common/coduno-iron-ajax.html:6]: replace `{{}}` with `[[]]` except for `last-response`. Add `auto="[[auto]]"`. Add a provider with the `baseUrl`(like in providers) and `url="[[baseUrl]]/[[path]]"`

File: app/elements/common/coduno-iron-ajax.html [hunk 4]
```diff
@@ -45,30 +43,34 @@
...
-					notify: true
-				},
-				debounceDuration: {
-					type: Number,
-					notify: true,
-					value: 0
 				}
 			},
 			generateRequest: function(){
+				console.log(this.suffix);
```
Comment (victorbalan) [inline app/elements/common/coduno-iron-ajax.html]: d

File: app/elements/common/coduno-iron-ajax.html [hunk 5]
```diff
@@ -45,30 +43,34 @@
...
 			},
 			_onResponse: function(r){
+				this.error = null;
+				this.response = r.detail.response;
 				this.fire('response', r);
 			},
-			_onError: function(err){
-				this.fire('error', err);
+			_onError: function(e){
+				this.error = {
```
Comment (victorbalan) [inline app/elements/common/coduno-iron-ajax.html]: encapsulate this into a behavior too. Maybe we won't use coduno-iron-ajax in some places.

File: app/elements/common/coduno-iron-ajax.html [hunk 6]
```diff
@@ -21,20 +20,19 @@
...
-					reflectToAttrbute: true
-				},
-				lastResponse: {
+				error: {
 					type: Object,
 					notify: true,
-					reflectToAttrbute: true
+					reflectToAttribute: true
+				},
+				suffix: {
```
Comment (victorbalan) [inline app/elements/common/coduno-iron-ajax.html]: rename to path

File: app/scripts/behaviors/providers/general-provider-behavior.js [hunk 7]
```diff
@@ -24,5 +33,24 @@ Behaviors.Providers.General = {
 		}
 
 		this.baseUrl = prefix;
+	},
+	onResponse: function(r){
```
Comment (victorbalan) [inline app/scripts/behaviors/providers/general-provider-behavior.js]: please strip this from here into a separate behavior

**Comment by victorbalan** [comment]
everytime before you push please do a `gulp`. You will have approx 4 errors now. Our travis build does not detect these and if this ends up on master it will fail.

File: app/elements/common/data-providers/test/tests-by-task-provider.html [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+		:host {
+			display: none;
+		}
+	</style>
+	<template>
+		<iron-ajax
+			id="request"
+			auto
+			method="GET"
+			url="[[baseUrl]]/tasks/[[task.id]]/tests"
```
Comment (victorbalan) [inline app/elements/common/data-providers/test/tests-by-task-provider.html]: you know what to do.

File: app/elements/admin/import-catcoder-input.html [hunk 9]
```diff
@@ -4,53 +4,60 @@
...
+			paper-button {
+				margin: 8px 4px ;
+			}
+		</style>
+		<coduno-iron-ajax
 			id="postGame"
 			method="POST"
-			with-credentials="true"
-			on-response="_onResponse"
+			body="{{body}}"
```
Comment (victorbalan) [inline app/elements/admin/import-catcoder-input.html]: `{{body}}` -> `[[body]]`

File: app/elements/coder/challenge/available/register-to-challenge-dialog.html [hunk 10]
```diff
@@ -86,20 +87,13 @@
...
 				this.$.registerRequest.generateRequest();
 			},
-			_onResponse: function(){
+			_onRequestFinished: function(e){
+				e.stopPropagation();
 				this.toastText = 'Successfuly registered.';
 				this.$.toast.show();
 				this.fire('challenge-registered');
-			},
-			_onError: function(e, r){
```
Comment (victorbalan) [inline app/elements/coder/challenge/available/register-to-challenge-dialog.html:97]: why remove the on error? What if the status is not 500 or 401 ? What if it's 409 `you are already invited to the challenge`

File: app/elements/coder/code-editor/consoles/tabbed-console.html [hunk 11]
```diff
@@ -27,15 +27,14 @@
...
-			handle-as="json"
-			on-response="onTestsResponse"
-			on-error="onTestsError"
-			method="GET"
-			with-credentials="true"
-			>
-		</iron-ajax>
+		<tests-by-task-provider
+			id="testsProvider"
+			task="{{task}}"
```
Comment (victorbalan) [inline app/elements/coder/code-editor/consoles/tabbed-console.html]: task -> `task-id="[[task.id]]"`

File: app/elements/coder/code-editor/consoles/tabbed-console.html [hunk 12]
```diff
@@ -27,15 +27,14 @@
...
-			id="getTestsRequest"
-			handle-as="json"
-			on-response="onTestsResponse"
-			on-error="onTestsError"
-			method="GET"
-			with-credentials="true"
-			>
-		</iron-ajax>
+		<tests-by-task-provider
+			id="testsProvider"
```
Comment (victorbalan) [inline app/elements/coder/code-editor/consoles/tabbed-console.html]: id not needed

File: app/elements/coder/code-editor/consoles/test-stats-console.html [hunk 13]
```diff
@@ -20,16 +20,16 @@
...
-			id="getTestsRequest"
-			handle-as="json"
-			on-response="onTestsResponse"
-			on-error="onTestsError"
-			method="GET"
-			with-credentials="true"
-			>
-		</iron-ajax>
+		<tests-by-task-provider
+			id="testsProvider"
```
Comment (victorbalan) [inline app/elements/coder/code-editor/consoles/test-stats-console.html]: same as above

File: app/elements/coder/input-view/input-view.html [hunk 14]
```diff
@@ -10,15 +11,13 @@
...
-			handle-as="json"
-			on-response="onTestsResponse"
-			on-error="onTestsError"
-			method="GET"
-			with-credentials="true"
-			>
-		</iron-ajax>
+		<tests-by-task-provider
+			id="testsProvider"
+			task="{{task}}"
```
Comment (victorbalan) [inline app/elements/coder/input-view/input-view.html]: same as above

File: app/elements/coder/team/create-team-modal.html [hunk 15]
```diff
@@ -17,18 +17,18 @@
...
+		<coduno-iron-ajax
 			id="createTeamRequest"
 			method="POST"
-			handle-as="text"
-			content-type="application/json"
-			with-credentials="true"
-			on-response="_onResponse"
-			on-error="_onError">
-		</iron-ajax>
+			body="{{team}}"
```
Comment (victorbalan) [inline app/elements/coder/team/create-team-modal.html]: same

File: app/elements/common/coduno-iron-ajax.html [hunk 16]
```diff
@@ -3,38 +3,31 @@
...
-		  handle-as="{{handleAs}}"
-		  last-response="{{lastResponse}}"
-		  headers="{{headers}}"
-		  url="{{url}}"
-			body="{{body}}"
-		  debounce-duration="{{debounceDuration}}"
-			on-response="_onResponse"
-			on-error="_onError"
-			with-credentials="true">
+			id="ajaxRequest"
```
Comment (victorbalan) [inline app/elements/common/coduno-iron-ajax.html:6]: add auto param

File: app/elements/coder/challenge/available/register-to-challenge-dialog.html (hunk 10 above)
Comment (pbochis) [inline app/elements/coder/challenge/available/register-to-challenge-dialog.html:97]: the error is handled by `coduno-iron-ajax` (handler behavior) and `errors-display` will show the error message from the platform.

File: app/elements/coder/team/create-team-modal.html [hunk 17]
```diff
@@ -60,32 +60,24 @@
...
+				this.$.createTeamDialog.open();
 			},
 			_saveTeam: function(){
 				if(!this.$.form.validate()){
 					return;
 				}
-				this.$.createTeamRequest.body = this.team;
-				this.$.createTeamRequest.url = util.build('/teams');
 				this.$.createTeamRequest.generateRequest();
 				this.team = {};
```
Comment (victorbalan) [inline app/elements/coder/team/create-team-modal.html]: make this empty when u get a response. If it is failed then maybe the user wants to modify what he wrote

**Comment by victorbalan** [comment]
http://prntscr.com/bb28x9 ugly space above title is from error detail empty

**Comment by victorbalan** [comment]
to make the dialogs work import them in elements.html

**Comment by victorbalan** [comment]
when u invite and the user is invalid the title is changed(no team name)

**Comment by victorbalan** [comment]
http://prntscr.com/bb2a5m on delete team

**Comment by victorbalan** [comment]
http://prntscr.com/bb2aea

**Comment by victorbalan** [comment]
http://prntscr.com/bb2asg

**Comment by pbochis** [comment]
The last screenshot is rly bad xD

**Pull request merged**